Keep indexed values' order

erral opened the pull request:
The taxonomy value indexer has to check all items in the taxonomy in order to index in the catalog also the parent items.

If your taxonomy is, as follows:

- 1
- 2
- - 2.1
- - 2.2
- 3
- 4
- - 4.1
- - 4.2

If you select `2` and `4.2`, the taxonomy indexer will add `2`, `4.2` and `4` to the catalog index and catalog metadata. So far so good.

To do so, the indexer uses an iterative approach, where it loops over all values of the taxonomy in the current language, and compares its titles splitting them with the PATH_SEPARATOR.

It works as expected, returning the required values.

But when returning the values, it doesn't keep the order in which the values were added to the content object.

In Classic UI the multiselect widget is shown as a multichoice widget with ordering. 

![2025-03-07 07-58-58](https://github.com/user-attachments/assets/70f5a0c1-93d5-40ea-86d8-0023988cd84b)

In Volto a element picker is used but you can reorder the items, or each time you select an item it goes to the end of the selection.

![2025-03-07 07-59-35](https://github.com/user-attachments/assets/ae838096-7784-43e0-a66f-0603f97e0ccc)

Our client, and our QA too, expected that the items shown in the page keep the same order as the selected.

This is achieved when you are showing the items in the content view (so getting the values from the object itself), but now when showing the items directly after a catalog query because in this case the values come from the catalog metadata (one can even argue if the "parent" items should ever be in the catalog metadata, but I leave that for another issue).

So, with these changes we keep the ordering when indexing the values.

I have also optimized a bit the indexer.

I have added a test with the usecase and all other test are still passing.

erral commented:
> Hm... first of all it sounds wrong to me, that when you select `4.2` the index `4` is also added ... I do not want all items on level 4 if I select 4.2 ...
> 

Well, this is how taxonomy index works, because it is supposed that `4.2` is also `4`, so if the user searches items with `4`. the item with `4.2` should be found.

> regarding ordering ... I'm not sure if I understand the issue here correctly, but does our issue (and PR) here https://github.com/collective/collective.taxonomy/issues/71 is eventually related?

Let me check this, because it could be related.

petschki commented:
> Well, this is how taxonomy index works, because it is supposed that `4.2` is also `4`, so if the user searches items with `4`. the item with `4.2` should be found.

of course that makes sense ... first I thought the other way round ... sorry.

petschki commented:
regarding tests: `plone/setup-plone@v3.0.0` should fix the setuptools problem https://github.com/collective/collective.taxonomy/blob/erral-indexing/.github/workflows/main.yml#L41

petschki commented:
Hm ... I see, that I got stuck too here https://github.com/collective/collective.taxonomy/actions/runs/12745480619/job/35519527036?pr=171 ... if you fix the testing issues I'll rebase my PR with the orderedselect widget ... btw. which Plone version are you using?

erral commented:
> Hm ... I see, that I got stuck too here https://github.com/collective/collective.taxonomy/actions/runs/12745480619/job/35519527036?pr=171 ... if you fix the testing issues I'll rebase my PR with the orderedselect widget ... btw. which Plone version are you using?

I am testing this in Plone 6.0.14

erral commented:
I have updated the plone-setup action but it fails in the Plone 6 and python 3.8  matrix because of setuptools :(

mamico commented:
> I have updated the plone-setup action but it fails in the Plone 6 and python 3.8 matrix because of setuptools :(

The problem is that setuptools 75.6.0, which is required in [Plone's constraints.txt](https://dist.plone.org/release/6.0-latest/constraints.txt), is not supported on Python 3.8 (see [setuptools 75.6.0 on PyPI](https://pypi.org/project/setuptools/75.6.0/)). And, if I remember well, Python 3.8 support is also removed from Plone 6.0.

A quick fix is to remove Python 3.8 from tests and from the list of supported versions for this package as well.

erral commented:
> > I have updated the plone-setup action but it fails in the Plone 6 and python 3.8 matrix because of setuptools :(
> 
> The problem is that setuptools 75.6.0, which is required in [Plone's constraints.txt](https://dist.plone.org/release/6.0-latest/constraints.txt), is not supported on Python 3.8 (see [setuptools 75.6.0 on PyPI](https://pypi.org/project/setuptools/75.6.0/)). And, if I remember well, Python 3.8 support is also removed from Plone 6.0.
> 
> A quick fix is to remove Python 3.8 from tests and from the list of supported versions for this package as well.

That worked, thanks!

1letter reviewed (approved):
LGTM, i run tox -e lint and pushed a small change.

erral commented:
> LGTM, i run tox -e lint and pushed a small change.

Thanks, I was stuck running `make lint` which does different things :/

petschki commented:
I've rebased this to the new `plone/meta=2.x` configuration ... the `test` and `qa` actions are now done by meta.
